JavaScript: Add new query flagging identity replacements.

xiemaisi opened the pull request:
This is a [surprisingly common](https://lgtm.com/query/1506552808324) mistake.

esben-semmle commented on javascript/ql/src/RegExp/IdentityReplacement.ql:
```diff
@@ -0,0 +1,44 @@
+ */
+predicate matchesString(Expr e, string s) {
+  exists (RegExpConstant c |
+    matchesConstant(e.(RegExpLiteral).getRoot(), c) and
```
We should exclude the `i`-flag for cases like: `x.replace(/foo/i, 'foo')`.

xiemaisi replied:
An excellent point, thanks!

esben-semmle commented on javascript/ql/src/RegExp/IdentityReplacement.ql:
```diff
@@ -0,0 +1,44 @@
+  exists (RegExpCharacterClass recc | recc = t and not recc.isInverted() |
+    recc.getNumChild() = 1 and
+    matchesConstant(recc.getChild(0), c)
+  )
```
We should handle anchors for cases like `/^x$/`

xiemaisi replied:
Good idea; and also other zero-width assertions like `\b`.

esben-semmle commented on javascript/ql/src/RegExp/IdentityReplacement.ql:
```diff
@@ -0,0 +1,44 @@
+ * @kind problem
+ * @problem.severity warning
+ * @id js/identity-replacement
+ * @tags correctness
```
Have you considered the `security` tag?

xiemaisi replied:
It's a bit of a stretch, but I suppose it does cover CWE-116, so might as well claim it.

xiemaisi commented:
@esben-semmle: Updated as suggested. I've amended the original commit to include the changes to the query metadata and added it to a suite, the other commits roughly correspond to your other suggestions.

This query is now a bit more complex. I could run a full performance evaluation if desired, but it still looks fairly snappy on the query console, and takes about 20 seconds on `gecko-dev`, which has some pretty gnarly regular expressions.

xiemaisi commented on javascript/ql/src/RegExp/IdentityReplacement.ql:
```diff
@@ -60,8 +60,9 @@ predicate regExpMatchesString(RegExpTerm t, string s) {
-      repl.getArgument(1).getStringValue() = s
-select repl.getArgument(0), "This replaces '" + s + "' with itself."
+      repl.getArgument(1).getStringValue() = s and
+      (if s = "" then friendly = "the empty string" else friendly = "'" + s + "'")
```
Could do, but once we start going down that route there are many other friendliness improvements we could also consider (such as special handling of newlines and whatnot). This one is very simple and straightforward, the other ones not so much.

xiemaisi commented on change-notes/1.19/analysis-javascript.md:
```diff
@@ -14,6 +14,7 @@
 |-----------------------------------------------|------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
 | Enabling Node.js integration for Electron web content renderers (`js/enabling-electron-renderer-node-integration`) | security, frameworks/electron, external/cwe/cwe-094  | Highlights Electron web content renderer preferences with Node.js integration enabled, indicating a violation of [CWE-94](https://cwe.mitre.org/data/definitions/94.html). Results are not shown on LGTM by default. |
 | Stored cross-site scripting (`js/stored-xss`) | security, external/cwe/cwe-079, external/cwe/cwe-116 | Highlights uncontrolled stored values flowing into HTML content, indicating a violation of [CWE-079](https://cwe.mitre.org/data/definitions/79.html). Results shown on LGTM by default. |
+| Replacement of a substring with itself (`js/identity-replacement`) | correctness | Highlights string replacements that replace a string with itself, which usually indicates a mistake. Results shown on LGTM by default. |
```
Ah, of course; fixed. I decided not to go for the violation-of-CWE phrasing, though, since I don't want to claim that that's the main purpose of the query.

xiemaisi commented:
Nits addressed/responded to.

esben-semmle commented:
Ping @mc-semmle for doc-review.
